frontend/src/metabase/visualizations/lib/chart_values.js

```diff
@@ -82,20 +83,38 @@ export function onRenderValueLabels(
       })
       .value();
 
-    return data
+    data = data
       .map(([x, y], i) => {
         const isLocalMin =
           // first point or prior is greater than y
           (i === 0 || data[i - 1][1] > y) &&
           // last point point or next is greater than y
           (i === data.length - 1 || data[i + 1][1] > y);
         const showLabelBelow = isLocalMin && display === "line";
-        const rotated = barCount > 1 && display === "bar" && barWidth < 40;
+        const rotated =
+          barCount > 1 &&
+          (display === "bar" || display === "waterfall") &&
+          barWidth < 40;
         const hidden =
-          !showAll && barCount > 1 && display === "bar" && barWidth < 20;
+          !showAll &&
+          barCount > 1 &&
+          (display === "bar" || display === "waterfall") &&
+          barWidth < 20;
         return { x, y, showLabelBelow, seriesIndex, rotated, hidden };
       })
-      .filter(d => !(display === "bar" && d.y === 0));
+      .filter(
+        d => !((display === "bar" || display === "waterfall") && d.y === 0),
+      );
+
+    if (display === "waterfall") {
+      let total = 0;
+      data.forEach(d => {
+        d.cumulativeY = d.y + total;
```

Comment:
This is probably a bad idea, but I wanted to float it anyways: could we use the beam values instead of recomputing cumulativeY here?

Reply:
Unfortunately, since this is at `onRenderValueLabels`, the additional series data is not handily available (or at least, I can't find a way to either access it or smuggle it easily).